### ISSUE TITLE
respond_to_missing? called on mock

### ISSUE THREAD

**Issue by t3hk0d3** [rspec/rspec-mocks#1066]
```ruby
class TestMe
  class << self

    def callme
      'failure'
    end

    private

    def method_missing(name, *args, &block)
      raise "Failure!"
    end

    def respond_to_missing?(name, include_private = false)
      callme.respond_to?(name, include_private) || super
    end
  end
end

require 'rspec'

describe TestMe do
  it 'mock static method' do
    allow(TestMe).to receive(:callme).and_return('success')
    expect(TestMe.callme).to eq('success')
  end
end
```

**Comment by myronmarston** [comment]
> FUNFACT: It worked well in RSpec 3.2, and broke only after update to 3.4


Thanks for mentioning this; it helped me bisect it back to the culprit -- 8ea8b6016d2b7cd70fef0a0c2687ba70ecf8adb5 (the change from #1047 for 3-4-maintenance).  To avoid some Ruby 2.3 warnings we fixed a bug in the `InstanceMethodStasher` that causes the stubbed method to be removed.  It's not obvious to me what, if anything, we can do to fix this -- or even if we should.

In general, `respond_to?` (and by extension, `respond_to_missing?`) is part of Ruby's meta-object protocol that rspec-mocks relies upon.  To provide the features of rspec-mocks we sometimes have to introduce new calls to it (this is what happened between 3.2 and 3.4).  We expect `respond_to?` to return `true` or `false` (or at least a truthy or falsey value).  An object that raises an exception from `respond_to?` is a mis-behaving object that we're not able to support.

Your case is a weird case, though; it looks like it shouldn't generally raise an error from `respond_to?`.  I'm not sure what the solution is yet.

**Comment by t3hk0d3** [comment]
@myronmarston It's not about exception. Its about stubbed method being called from ``respond_to_missing?`` being routed to ``method_missing``. Exception is only for indication and backtrace.

**Comment by myronmarston** [comment]
Can you provide another example showing the problem w/o the exception?

**Comment by JonRowe** [comment]
`callme` is being called here and triggering method_missing

**Comment by t3hk0d3** [comment]
@JonRowe exactly! That could be happening because real method is already undefined (renamed) at moment ``respond_to_missing?`` is called.

Also i've tested that spec on 3.4.1 and it works as intended, but fail on 3.4.4.

@myronmarston 
What the point? Exception here only for indication ``method_missing`` being called.
